src/main/scala/com/microsoft/hyperspace/index/IndexLogEntry.scala

```diff
@@ -493,3 +541,50 @@ object IndexLogEntry {
 
   def schemaString(schema: StructType): String = schema.json
 }
+
+/**
+ * Provides functionality to generate unqiue file ids for source data files
+ * and track them.
+ *
+ * @param lastId Last assigned id. New ids will be greater than this value.
+ */
+class FileIdTracker(private val lastId: Long) {
+  private var maxId: Long = lastId
+  private val fileNameToIdMap: mutable.HashMap[String, Long] = mutable.HashMap()
+
+  def getMaxFileId: Long = maxId
+
+  def getFileNameToIdMap: HashMap[String, Long] = fileNameToIdMap
+
+  def getFileId(filePath: String): Option[Long] = fileNameToIdMap.get(filePath)
+
+  def setSizeHint(size: Int): Unit = fileNameToIdMap.sizeHint(size)
+
+  def addFileIds(map: HashMap[String, Long]): Unit = fileNameToIdMap ++= map
+
+  /**
+   * Try to add file path to fileNameToIdMap. If the file is already in
+   * the map then return its id. Otherwise, generate a new id, according
+   * to current maxId, and update fileNameToIdMap.
+   *
+   * @param file FileStatus to be lookedup in fileNameToIdMap.
+   * @return assigned id to given file.
+   */
+  def addFile(file: FileStatus): Long = {
+    val fullPath = file.getPath.toString
+    fileNameToIdMap.getOrElseUpdate(
+      fullPath,
+      {
+        maxId += 1
+        maxId
+      })
+  }
+}
+
+object FileIdTracker {
+  def apply(lastId: Long, map: HashMap[String, Long]): FileIdTracker = {
```

Comment:
Why not introduce a "add*" def with a proper comment of what `HashMap[String, Long]` is?

Reply:
Can you explain your comment here? Are you against adding object/apply for FileIdTracker?

Reply:
Oh I misread. For some reason, I thought this was adding new type of data structures. Looking at the usage, this is called from:
```
val fileNameToIdMap = mutable.HashMap[String, Long]()
sourceFileInfoSet.foreach(f => fileNameToIdMap.put(f.name, f.id))
```
Can't you just create add def that takes in `sourceFileInfoSet`?

Reply:
So are you suggesting below?
1- Drop the object/apply implementation
2- Add an addXX() function to the *class* to populate the Map using a Seq[FileInfo]
3- Changes the usage so it creates an empty FileIdTracker and then populate its map using the func from (2)

Reply:
right. we don't need to create `mutable.HashMap[String, Long]()` since the file tracker should be the one doing it.

Reply:
Oh I see what you are doing here. You need to maintain the id right?

Reply:
Ah, in this case, let's introduce a constructor in `FileIdTracker` that takes in `sourceFileInfoSet` and populates the mapping.

Reply:
Something like the following:
```scala
class FileIdTracker {                       // <= I don't think we need to expose startId anymore.
  private var maxId: Long = -1
  private val fileNameToIdMap: mutable.HashMap[String, Long] = mutable.HashMap()

  // FileInfo's 'name' contains the full path to the file.
  def this(files: Set[FileInfo]) = {
    this()
    files.foreach { f =>
      fileNameToIdMap.put(f.name, f.id)
      maxId = max(maxId, f.id)
    }
  }
```

The reason `def addFileIds(map: HashMap[String, Long]): Unit = fileNameToIdMap ++= map` is weird is that we need to handle the case where keys are the same but the values are different. With the constructor approach we don't need to worry about it anymore.

Reply:
@sezruby also suggested exposing add* def instead of constructor. I am also fine with that as long as we make sure there does not exist a case where there are different ids with the same key.